CovidWatch/libcontactrace/src/main/java/org/covidwatch/libcontactrace/BluetoothService.kt

```diff
@@ -0,0 +1,104 @@
+package org.covidwatch.libcontactrace
+
+import android.app.Notification
+import android.bluetooth.BluetoothAdapter
+import android.content.Intent
+import android.os.Binder
+import android.os.IBinder
+import androidx.lifecycle.LifecycleService
+import org.covidwatch.libcontactrace.cen.CenGenerator
+import org.covidwatch.libcontactrace.cen.CenVisitor
+import java.util.*
+
+class BluetoothService : LifecycleService() {
+
+    private var config: ServiceConfiguration? = null
+    private var cenAdvertiser: CenAdvertiser? = null
+    private var cenScanner: CenScanner? = null
+    private val binder: IBinder = LocalBinder()
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        super.onStartCommand(intent, flags, startId)
+        return START_STICKY
+    }
+
+    override fun onDestroy() {
+        cenAdvertiser?.stopAdvertiser()
+        cenScanner?.stopScanning()
+        super.onDestroy()
+    }
+
+
+    override fun onBind(intent: Intent): IBinder? {
+        super.onBind(intent)
+        return binder
+    }
+
+    private fun BluetoothAdapter.supportsAdvertising() =
+        isMultipleAdvertisementSupported && bluetoothLeAdvertiser != null
+
+    fun start() {
+        config?.apply {
+            startForeground(NOTIFICATION_ID, foregroundNotification)
+            val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
+
+            // advertise CEN
+            cenAdvertiser = CenAdvertiser(
+                this@BluetoothService,
+                bluetoothAdapter.bluetoothLeAdvertiser,
+                CONTACT_EVENT_SERVICE,
+                CONTACT_EVENT_IDENTIFIER_CHARACTERISTIC,
+                cenVisitor,
+                cenGenerator
+            )
+
+            // scan CENs
+            cenScanner = CenScanner(
+                this@BluetoothService,
+                bluetoothAdapter.bluetoothLeScanner,
+                CONTACT_EVENT_SERVICE,
+                cenVisitor
+            )
+            cenAdvertiser?.startAdvertiser(CONTACT_EVENT_SERVICE)
+            cenScanner?.startScanning(arrayOf(CONTACT_EVENT_SERVICE), 10)
+        }
+    }
+
+    fun configure(configuration: ServiceConfiguration) {
+        this.config = configuration
+    }
+
+    fun updateCen() {
```

Comment:
where is this invoked from?

Reply:
From the `timer` in `BluetoothManager` class.